Provide json_timestamp_layout option

HeikoSchlittermann opened the pull request:
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [ ] Has appropriate unit tests.

HeikoSchlittermann commented:
I'm updating the json_test.go right now, and run the tests locally. I'll push as soon as possible.

HeikoSchlittermann commented:
Is there any chance to get this PR merged into the upstream sources? It would greatly improve the chance for the integration of telegraf into a legacy environment.

sspaink reviewed (changes requested):
Thank you for the pull request, added two comments for you to review. Overall it looks good I think. I know its been a while since you've made this pr, are you still using this change?

HeikoSchlittermann commented on plugins/serializers/json/README.md:
```diff
@@ -19,6 +19,13 @@ The `json` output data format converts metrics into JSON documents.
+  # Other timestamp layout can be configured using the Go language time
+  # layout specification from https://golang.org/pkg/time/#Time.Format
+  # e.g.: json_timestamp_layout = "2006-01-02T15:04:05Z07:00"
+  #json_timestamp_layout = ""
```
Yes, we're still using it. Renaming the option shouldn't be a major challenge, but rebasing out patch onto the latest master is challenging on `config/*go`.

HeikoSchlittermann replied:
The updated PR now uses json.timestamp_format as suggested.

HeikoSchlittermann commented on plugins/serializers/json/json.go:
```diff
@@ -71,7 +73,11 @@ func (s *serializer) createObject(metric telegraf.Metric) map[string]interface{}
+	if s.TimestampLayout == "" {
+		m["timestamp"] = metric.Time().UnixNano() / int64(s.TimestampUnits)
+	} else {
+		m["timestamp"] = metric.Time().UTC().Format(s.TimestampLayout)
```
Hm, maybe I got something wrong, but ParseTimestamp converts from a _whatever_ format into `time.Time`, while the `json.TimestampLayout` / `json.TimestampFormat` deals with the oppisite way, from internal representation to a _string_ in the requsted format

HeikoSchlittermann commented:
Please apply this before #8184. I got a circular dependency in the unit tests, which I can fix as soon as *this* (#8229) is merged into master.

sspaink reviewed (approved):
lgtm, looks like there are some linter errors that will need to be fixed first before merging.

HeikoSchlittermann commented:
The linter errors are to be outside of this patche's scope, aren't they.

sspaink commented:
The linter error `var-naming: func NewJsonSerializer should be NewJSONSerializer` is in scope because that function name was changed in this pr. 

The other linter error could be considered out of the scope `unexported-return` but it is just capitalizing the serializer struct, I could do it in a separate pr but then you would have to rebase I assume? Not sure what would be best.

HeikoSchlittermann commented:
Hi Sebastian,
I'm about to leave for a 10 days vacation trip right now, so please
expect some delay in my reaction.

And thank you for caring about my PRs.
-- 
Heiko

sspaink commented on plugins/outputs/azure_data_explorer/azure_data_explorer.go:
```diff
@@ -230,7 +230,7 @@ func (adx *AzureDataExplorer) Init() error {
 	}
 
-	serializer, err := json.NewSerializer(time.Second)
+	serializer, err := json.NewSerializer(time.Second, "") // FIXME: get the json.TimestampFormat from the config file
```
Just noticed this, seems like this should be resolved before we merge this PR so that it can be used in this plugin? What plugin are you planning on using `json_timestamp_layout`?

HeikoSchlittermann replied:
The `http` output plugin is the one we use the `json_timestamp_format` (or "layout") with. But, as the *azure* output plugin seems to make use of the Json serializer too, I had to add the additional (empty) parameter.

HeikoSchlittermann replied:
Is there anything I should do here? Or is this explanation sufficient?

HeikoSchlittermann replied:
ping

HeikoSchlittermann commented:
What's the next step? I'm asking, as, once it is merged, I'd care about my other PR (http unbatched output).

sspaink commented on plugins/serializers/registry.go:
```diff
@@ -188,8 +191,8 @@ func NewWavefrontSerializer(prefix string, useStrict bool, sourceOverride []stri
 
-func NewJSONSerializer(timestampUnits time.Duration) (Serializer, error) {
-	return json.NewSerializer(timestampUnits)
+func NewJsonSerializer(timestampUnits time.Duration, timestampFormat string) (Serializer, error) {
```
Can you undo this capitalization change to appease the linter?
```suggestion
func NewJSONSerializer(timestampUnits time.Duration, timestampFormat string) (Serializer, error) {
```

HeikoSchlittermann replied:
Ah, thanks, I was too stupid to grasp the idea of the linter.

sspaink commented on plugins/serializers/registry.go:
```diff
@@ -123,7 +126,7 @@ func NewSerializer(config *Config) (Serializer, error) {
 		serializer, err = NewGraphiteSerializer(config.Prefix, config.Template, config.GraphiteTagSupport, config.GraphiteTagSanitizeMode, config.GraphiteSeparator, config.Templates)
 	case "json":
-		serializer, err = NewJSONSerializer(config.TimestampUnits)
+		serializer, err = NewJsonSerializer(config.TimestampUnits, config.TimestampFormat)
```
```suggestion
		serializer, err = NewJSONSerializer(config.TimestampUnits, config.TimestampFormat)
```

sspaink reviewed (changes requested):
Requesting a minor change for linter, then next steps would be someone else on the team needs to review this PR as well and then it can get merged.

HeikoSchlittermann commented:
Hello, is there something I can do to get the changes approved?

sspaink commented:
!retry-checks

popey commented:
Apologies for the delay in getting to this PR. It looks like CI is failing because we've updated the required Go version from 1.16.6 to 1.17 since you made this PR. Please could you rebase off master and once CI passes we'd be happy to merge it. Thanks!

HeikoSchlittermann commented:
Hi @popey , I rebased (and squashed) my changes onto the head of the current master. Please review again and tell me what changes are necessary to get it upstream.

Thanks and Greetings from Dresden/Germany.

HeikoSchlittermann commented:
Right now I changed the commit message to hopefully make the semantic pull request checker happy. And I seem to have succeeded :)
